examples/cloudml-churn-prediction/preprocessor/preprocessor/preprocess.py

```diff
@@ -0,0 +1,342 @@
+        default='bigquery-public-data.google_analytics_sample.ga_sessions_*',
+    )
+    parser.add_argument(
+        '--machine_type',
+        help="""Set machine type for Dataflow worker machines.""",
+        default='n1-highmem-4',
+    )
+    parser.add_argument(
+        '--cloud',
+        help="""Run preprocessing on the cloud. Default False.""",
+        action='store_true',
+        default=False,
+    )
+    parser.add_argument(
+        '--project_id',
+        help="""Google Cloud project ID""",
+        required=True,
+    )
+    known_args, _ = parser.parse_known_args(argv)
+
+    return known_args
+
+
+def get_pipeline_args(flags):
+    """Create Apache Beam pipeline arguments."""
+
+    options = {
+        'project': flags.project_id,
+        'staging_location': os.path.join(flags.output_dir, 'staging'),
+        'temp_location': os.path.join(flags.output_dir, 'temp2'),
+        'job_name': flags.job_name,
+        'save_main_session': True,
+        'setup_file': './setup.py'
+    }
+    return options
+
+
+# pylint: disable=expression-not-assigned
+def build_pipeline(p, flags):
+    """Sets up Apache Beam pipeline for execution."""
+
+    raw_data = (
+        p | 'QueryTable' >> beam.io.Read(
+            beam.io.BigQuerySource(
+                query=query.get_query(flags.bq_table),
+                project=flags.project_id,
+                use_standard_sql=True)
+            )
+        # omit 'Generate Data' step if working with real data
+        | 'Generate Data' >> beam.Map(_generate_fake_data)
+        | 'Extract duration and Label' >> beam.Map(_map_to_class)
+        | 'Generate label array' >> beam.Map(_combine_censorship_duration)
+        )
+    raw_train, raw_eval, raw_test = (
+        raw_data | 'RandomlySplitData' >> randomly_split(
+            train_size=.7,
+            validation_size=.15,
+            test_size=.15))
+    raw_metadata = features.get_raw_dataset_metadata()
+    preprocess_fn = features.preprocess_fn
+    transform_fn = (
+        (raw_train, raw_metadata) | 'AnalyzeTrain' >> tft_beam.AnalyzeDataset(
+            preprocess_fn))
+    (transform_fn | 'WriteTransformFn' >> tft_beam_io.WriteTransformFn(
+        flags.output_dir))
+
+    for dataset_type, dataset in [('Train', raw_train),
+                                  ('Eval', raw_eval),
+                                  ('Test', raw_test)]:
+        transform_label = 'Transform{}'.format(dataset_type)
+        t, metadata = (((dataset, raw_metadata), transform_fn)
+                       | transform_label >> tft_beam.TransformDataset())
+        if dataset_type == 'Train':
+            (metadata | 'WriteMetadata' >> tft_beam_io.WriteMetadata(
+                os.path.join(
+                    flags.output_dir, 'transformed_metadata'), pipeline=p))
+        write_label = 'Write{}TFRecord'.format(dataset_type)
+        t | write_label >> write_tfrecord(
+            dataset_type, flags.output_dir, metadata)
+
+
+def run():
```

Comment:
This is confusing as `beam.Pipeline` has a `run` method.
In fact, AFAICT this method doesn't seem to run the pipeline at all, it just constructs a pipeline object. 
I believe you need a to call `p.run()` (assuming is an instance of `beam.Pipeline`)

Reply:
I don't think p.run() is required if you're using it with "with beam.Pipeline(runner, options=options) as p:". I've at least never have seen that in any code samples. 

Would it be less confusing if I renamed this function? (although using run() also follows the Juliaset example)